sotodlib/io/hk_utils.py

```diff
@@ -0,0 +1,133 @@
+
+    grouped_feeds = []
+    key_func = lambda field: field.split('.')[0:4]
+    for key, group in itertools.groupby(sorted_fields, key_func):
+        grouped_feeds.append(list(group))
+
+    # use grouped_feeds and fields_data to output a grouped list of data per
+    # feed/device, ready to input to an axismanager
+    grouped_data = []
+    for group in grouped_feeds:
+        device_data = {}
+        for field in group:
+            field_dict = {field: fields_data[field]}
+            device_data.update(field_dict)
+
+        grouped_data.append(device_data)
+
+    return grouped_data
+
+
+def make_hkaman(grouped_data):
+    """
+    Takes data from get_grouped_hkdata(), tests whether feed/device is
+    cosampled, outputs axismanager(s) for either case.
+    """
+    amans = []
+    for group in grouped_data:
+        data = {}
+        aliases = []
+        times = []
+        for field in group:
+            # arrange data per device to input into aman
+            alias = group[field][0]
+            aliases.append(alias)
+
+            time = group[field][1]
+            times.append(time)
+
+            device_data = group[field][2]
+
+            info = {alias: device_data}
+            data.update(info)
+
+        try:
+            # check whether hk device is cosampled
+            assert len(times[0]) == len(times[-1])
+            # TODO: diff of times, if less than say 10%??, treat as cosampled
+        except AssertionError:
+            # if not cosampled, make aman for each field in device/feed
+            print("{} is not cosampled".format('.'.join(field.split('.')[0:3])))
+            for field in group:
+                alias = group[field][0]
+                time = group[field][1]
+                data = group[field][2]
+
+                device_name = field.split('.')[1]
+                device_axis = 'hkdata_' + device_name
+                samps_axis = 'hksamps_' + device_name
+
+                hkaman = core.AxisManager(core.LabelAxis(device_axis, [alias]),
+                                          core.OffsetAxis(samps_axis, len(time)))
+                hkaman.wrap('times', time, [(0, samps_axis)])
+                hkaman.wrap(device_name, np.array([data]), [(0, device_axis), (1, samps_axis)])
+                amans.append(hkaman)
+        else:
+            # if yes, make one aman per device
+            device_name = field.split('.')[1]
+            device_axis = 'hkdata_' + device_name
+            samps_axis = 'hksamps_' + device_name
+
+            hkaman_cos = core.AxisManager(core.LabelAxis(device_axis, aliases),
+                                          core.OffsetAxis(samps_axis, len(time)))
+            hkaman_cos.wrap('times', time, [(0, samps_axis)])
+            hkaman_cos.wrap(device_name, np.array([data[alias] for alias in aliases]), [(0, device_axis), (1, samps_axis)])
+            amans.append(hkaman_cos)
+
+        # TODO: make an aman of amans
```

Comment:
It looks like this should just be a series of `merge` or `wrap` commands. Have you tried it? I think it's worth just adding that in before merging.